dd-sdk-android-ndk/src/main/cpp/utils/signal-monitor.c

```diff
@@ -152,61 +146,127 @@ bool configure_signal_stack() {
     return true;
 }
 
-void invoke_previous_handler(int signum, siginfo_t *info,
-                             void *user_context) {
 
-    pthread_mutex_lock(&handlerMutex);
+bool configure_global_sigaction() {
+    global_sigaction = calloc(handled_signals_size(), sizeof(struct sigaction));
+    if (global_sigaction == NULL) {
+        return false;
+    }
+    sigemptyset(&global_sigaction->sa_mask);
+    global_sigaction->sa_sigaction = handle_signal;
+    // we will use the SA_ONSTACK mask here to handle the signal in a freshly new stack.
+    global_sigaction->sa_flags = SA_SIGINFO | SA_ONSTACK;
 
+    return true;
+}
+
+void *override_native_signal_handlers() {
+    pthread_mutex_lock(&mutex);
+    if (handlers_installed) {
+        pthread_mutex_unlock(&mutex);
+        pthread_exit(NULL);
+    }
     const size_t signals_array_size = handled_signals_size();
-    for (int i = 0; i < signals_array_size; ++i) {
+    global_previous_sigaction = calloc(signals_array_size, sizeof(struct sigaction));
+    if (global_previous_sigaction == NULL) {
+        __android_log_write(ANDROID_LOG_ERROR, LOG_TAG, "Was not able to initialise.");
+        pthread_mutex_unlock(&mutex);
+        pthread_exit(NULL);
+    }
+    for (int i = 0; i < signals_array_size; i++) {
         const int signal = handled_signals[i].signal_value;
-        if (signal == signum) {
-            struct sigaction previous = global_previous_sigaction[i];
-            // From sigaction(2):
-            //  > If act is non-zero, it specifies an action (SIG_DFL, SIG_IGN, or a
-            //  handler routine)
-            if (previous.sa_flags & SA_SIGINFO) {
-                // This handler can handle signal number, info, and user context
-                // (POSIX). From sigaction(2): > If this bit is set, the handler
-                // function is assumed to be pointed to by the sa_sigaction member of
-                // struct sigaction and should match the proto- type shown above or as
-                // below in EXAMPLES.  This bit should not be set when assigning SIG_DFL
-                // or SIG_IGN.
-                previous.sa_sigaction(signum, info, user_context);
-            } else if (previous.sa_handler == SIG_DFL) {
-                // raise to trigger the default handler. It cannot be called directly.
-                raise(signum);
+        int success = sigaction(signal, global_sigaction,
+                                &global_previous_sigaction[i]);
+        if (success != 0) {
+            __android_log_print(ANDROID_LOG_ERROR,
+                                LOG_TAG,
+                                "Was not able to catch the signal: %d",
+                                signal);
```

Comment:
return an error? do not set handlers_installed to true if they were actually not installed?

Reply:
Actually it can happen maybe that one of the handlers is not added correctly but the others are. That's the reason I set the `handlers_installed` to true to be able to restore the installed handlers later.